packages/rest/src/router/metadata.ts

```diff
@@ -72,10 +72,65 @@ export function api(spec: ControllerSpec) {
       'The @api decorator can be applied to constructors only.',
     );
     const apiSpec = resolveControllerSpec(constructor, spec);
-    Reflector.defineMetadata(API_SPEC_KEY, apiSpec, constructor);
+    const key = constructor.prototype._server
+      ? `${API_SPEC_KEY}:${constructor.prototype._server}`
+      : API_SPEC_KEY;
+    Reflector.defineMetadata(key, apiSpec, constructor);
   };
 }
 
+/**
+ * A decorator to mark a controller as belonging to a specific server.
+ * ```ts
+ * // In your application...
+ * class MyApplication extends Application {
+ *   constructor() {
+ *     super();
+ *     this.server(RestServer, 'foo');
+ *   }
+ * }
+ *
+ * // In one of your controllers
+ * @server('foo')
+ * class MyController {
+ *  // ...
+ * }
+ * ```
+ * @param name The name of the server bound to your application context.
+ */
+export function server(name: string) {
+  return function(constructor: Constructor<any>) {
+    const endpoints = Reflector.getOwnMetadata(ENDPOINTS_KEY, constructor);
+    if (endpoints) {
+      rescopeMetadata(ENDPOINTS_KEY, name, constructor, endpoints);
+    }
+    const spec = getControllerSpec(constructor);
+    if (spec) {
+      rescopeMetadata(API_SPEC_KEY, name, constructor, spec);
+    }
+    constructor.prototype._server = name;
+  };
+}
+
+/**
+ * Replace an existing metadata definition with a scoped version.
+ * @param key The key for the metadata (if it already exists).
+ * @param scope The new scope to constrain the metadata definition.
+ * @param constructor The constructor for which the metadata is defined.
+ * @param value The metadata itself.
+ */
+function rescopeMetadata(
```

Comment:
This function seems to be cryptic. I don't think we should add magic to the metadata key. If we need to access api specs for a given server, we should build a helper function such as `getControllerSpec(controllerClass, server)`.

Reply:
Yes, but how will you know which spec to retrieve? The purpose of tainting the existing key with a `:${server}` suffix is to scope that spec to the particular server.

Reply:
Well, we can get the full spec for the controller and filter on the `servers` attributes.

Reply:
We cannot expect decorators to fully resolve the metadata during `decoration`. There should be post-processing if needed to reconcile/aggregate metadata contributed by multiple decorators.

Reply:
I'm not sure what you mean here. If metadata is being defined inside the controller class (on individual methods) then it's all done before the `@servers` decorator is resolved.

The `@api` and `@servers` decorators could conflict, but only if you call the `@api` decorator with a specification that doesn't match any endpoints generated beforehand using method-level decorators (and we don't support this use-case; if you use the method-level decorators, you should avoid using the `@api` decorator)

Reply:
Decorators can be declared on a class by any order, for example:
- Case 1:
```
@api(...)
@servers(...)
class MyController {}
```

- Case 2:
```
Case 2:
@api(...)
@servers(...)
class MyController {}
```

In case 1, `servers() is executed before `api()`.

Reply:
@raymondfeng Yes, I understand that, but the only time where the order matters in this particular case is with the inner decorators used on class methods. If you're using the `@api` decorator to setup your spec, then the `@servers` decorator doesn't need to change anything since those specifications are controller-specific, not server-specific.